### PR TITLE
UNI-438 Bugfix: skipping many raw data points when merging with aggregated

### PR DESCRIPTION
**Pull request by mrcslws**
The issue here was in `_overlayDataSeries`. It's using a `.forEach` and not always doing something with the `item`.

I already reworked the way we build up this data for another change: https://github.com/numenta/numenta-apps/pull/761 . It fixes the bug. We're still discussing that change, so this PR is just that one without the controversial parts.

### CODE REVIEW

File: unicorn/app/browser/components/Chart.jsx [hunk 1]
```diff
@@ -138,7 +138,7 @@ export default class Chart extends React.Component {
     // init, render, and draw chart!
     options.labelsUTC = true;
     options.dateWindow = this._chartRange;  // update viewport of range selector
-    options.axes.y.valueRange = [metaData.min, metaData.max];  // lock y-axis
+    options.axes.y.valueRange = [metaData.min, metaData.max];
```
Comment (marionleborgne) [inline unicorn/app/browser/components/Chart.jsx]: why did you remove this comment?

File: unicorn/app/browser/components/ModelData.jsx [hunk 2]
```diff
@@ -38,6 +38,132 @@ const {
...
+  }
+
+  for (; j < b.length; j++) {
+    ret.push(b[j]);
+  }
+
+  return ret;
+}
+
+/**
```
Comment (marionleborgne) [inline unicorn/app/browser/components/ModelData.jsx]: why do you keep the heuristic here if it does nothing? is this to prepare for your other related PR https://github.com/numenta/numenta-apps/pull/761?

File: unicorn/app/browser/components/ModelData.jsx [hunk 3]
```diff
@@ -38,6 +38,132 @@ const {
...
+ * Use a heuristic to detect gaps in timestamps in the data.
+ *
+ * At each gap, insert [midpointOfGap, vals[0], vals[1], ...] as a new datum.
+ *
+ * @param {Array} data - Array of arrays: [[Date, ...], [Date, ...], ...]
+ * @param {Array} vals - Values concatenated to timestamp at every single gap
+ * @returns {Array} - data with gap values inserted
+ */
+function insertIntoGaps(data, vals) {
```
Comment (marionleborgne) [inline unicorn/app/browser/components/ModelData.jsx]: same question as above for this function

File: unicorn/app/browser/components/ModelData.jsx [hunk 4]
```diff
@@ -38,6 +38,132 @@ const {
...
+ * @param {boolean} aggregated - Whether the model is aggregated.
+ * @param {boolean} rawDataInBackground - Whether the raw data should be drawn
+ *                                        (in addition to the aggregated data)
+ * @returns {Array} - Tuple:
+ *                    Array: Dygraphs multi-dimensional array
+ *                    boolean: minVal
+ *                    boolean: maxVal
+ * @see http://dygraphs.com/tests/independent-series.html
+ */
+function prepareData(
```
Comment (marionleborgne) [inline unicorn/app/browser/components/ModelData.jsx]: 👍 nice refactoring. it fixes the bug of the missing points.

**Comment by marionleborgne** [comment]
👍 LGTM. If you could just reply to my two questions above, you can merge. It fixes the bug. Tests are passing (minus the one that's still failing on master - still working on that)

File: unicorn/app/browser/components/ModelData.jsx (hunk 2 above)
Comment (mrcslws) [inline unicorn/app/browser/components/ModelData.jsx]: Well, this `sortedMerge` function is used independently of the gap detection. It's used for merging the aggregated and raw data.

File: unicorn/app/browser/components/ModelData.jsx (hunk 3 above)
Comment (marionleborgne) [inline unicorn/app/browser/components/ModelData.jsx]: No, I am talking about the function `insertIntoGaps`

File: unicorn/app/browser/components/ModelData.jsx (hunk 2 above)
Comment (marionleborgne) [inline unicorn/app/browser/components/ModelData.jsx]: No, I am talking about the function `insertIntoGaps`

File: unicorn/app/browser/components/ModelData.jsx (hunk 3 above)
Comment (mrcslws) [inline unicorn/app/browser/components/ModelData.jsx]: I don't have a strong opinion here. I could totally remove it. It's just a nice placeholder to keep this code compatible with the gap detection while we figure out the actual heuristic.

File: unicorn/app/browser/components/Chart.jsx (hunk 1 above)
Comment (mrcslws) [inline unicorn/app/browser/components/Chart.jsx]: It's now updated on every chart update. Though yes it's still locked in some sense -- it doesn't shift with the selected timestep range. There's no right place to describe all of this in a single comment, and it just seems like this is all communicated by the code, no comment needed.

File: unicorn/app/browser/components/ModelData.jsx (hunk 2 above)
Comment (mrcslws) [inline unicorn/app/browser/components/ModelData.jsx]: Ahh. Same answer.

File: unicorn/app/browser/components/ModelData.jsx (hunk 3 above)
Comment (marionleborgne) [inline unicorn/app/browser/components/ModelData.jsx]: ok. we need the heuristic to detect gaps anyway so let's keep it.

**Comment by marionleborgne** [comment]
👍
